### ISSUE TITLE
[Training] Degenerate behavior when fine-tuning on official data

### ISSUE THREAD

**Issue by hectorcarrion** [NIRVANALAN/STream3R#17]
Hello Authors,

I am having some issues fine-tuning your model, as a sanity check I followed the official instructions to pre-process the MVS-Synth dataset. Afterward, I trained for one epoch on this data with the following config:

```
# @package _global_

defaults:
  - override /model: stream3r

seed: 42

data:
  data_scaling: 1.0
  data_root: /workspace/data/mvs-synth/processed_GTAV_540
  num_views: 24
  allow_repeat: true
  n_corres_train: 0

  resolution:
    - [224, 224]

  data_module:
    _target_: stream3r.data.multiview_dust3r_datamodule.MultiViewDUSt3RDataModule
    pin_memory: true
    num_workers: 16
    num_workers_val: 1
    batch_size_per_device: 1
    batch_size_per_device_val: 1

    train_datasets:
      - 1400 @ MVS_Synth_Multi(allow_repeat=${data.allow_repeat}, split='train', ROOT="${data.data_root}", aug_crop=16, resolution=${data.resolution}, transform=ColorJitter, num_views=${data.num_views}, n_corres=${data.n_corres_train})

model:
  pretrained: /workspace/models/base/model.safetensors

  scheduler:
    warmup_start_lr: 1e-7
    warmup_epochs: 2

  train_criterion:
    _target_: stream3r.loss.losses.CausalLoss
    gradient_loss: grad
    is_metric: false

  validation_criterion:
    _target_: stream3r.loss.losses.CausalLoss
    gradient_loss: grad
    is_metric: false

  optimizer:
    _target_: torch.optim.AdamW
    _partial_: true
    lr: 1e-6  # Lower LR when training encoder (10× lower than paper)
    betas:
      - 0.9
      - 0.95
    weight_decay: 0.05

trainer:
  devices: auto
  max_epochs: 50
  accumulate_grad_batches: 16  # Effective batch = 1 * 16 * 8 = 128 (larger for stability)
  strategy:
    _target_: lightning.pytorch.strategies.DeepSpeedStrategy
    timeout:
      _target_: datetime.timedelta
      minutes: 80
  plugins: null
  precision: bf16-mixed
  log_every_n_steps: 20
  limit_val_batches: 0

callbacks:
  model_checkpoint:
    every_n_train_steps: null
    every_n_epochs: 1
    save_top_k: 5
    filename: "{epoch:03d}-{step:08d}"
    save_last: false
    monitor: "train/loss"
    mode: "min"
```

As you can see I have very low LR and two steps of warmup. However, I observe the following inference result when using the resulting checkpoint on demo data:
<img width="1050" height="823" alt="Image" src="https://github.com/user-attachments/assets/a40a01b4-6731-48e3-ad90-a305c0605d97" />

Is this a behavior you observed before? Any clues would be very helpful, I'm happy to provide any additional information and help debug. Thanks!

**Comment by Luo-Yihang** [comment]
Hi! The results looks weired.

May I check the following:

1. What does the training loss curve look like?

2. How did you convert the trained checkpoint back into a .safetensors file for use in the demo?
Here is the script I used for your reference:

``` python
def convert_ckpt_to_safetensors(ckpt_path, output_path):
    print(f"Loading checkpoint from {ckpt_path}")
    ckpt = torch.load(ckpt_path, map_location="cpu", weights_only=False)
    
    if "state_dict" in ckpt:
        state_dict = ckpt["state_dict"]
    else:
        state_dict = ckpt
    
    print("Converting state dict...")
    converted_state_dict = {}
    for k, v in state_dict.items():
        if k.startswith("net."):
            k = k[4:]  # Remove "net." prefix
        converted_state_dict[k] = v.contiguous()
   
    print(f"Saving to {output_path}")
    save_file(converted_state_dict, output_path)
    print("Done!")
```

**Comment by hectorcarrion** [comment]
Thanks for the superfast response!

1. The degeneration happens after a few steps, but its still there after long training runs, here are the curves:

Train:
<img width="1051" height="662" alt="Image" src="https://github.com/user-attachments/assets/02e76182-48db-4458-b681-fa59242c36fb" />

Val:
<img width="1058" height="659" alt="Image" src="https://github.com/user-attachments/assets/10c98116-b194-41a5-8efd-1f8061b0fb40" />

2. I downloaded the .safetensor from your HF repo, and I'm loading it for training this way:

```
    def _load_pretrained_weights(self) -> None:
        log.info(f"Loading pretrained: {self.pretrained}")
        if isinstance(self.net, FlashDUSt3R):
            ckpt = torch.load(self.pretrained)
            ckpt = self._update_ckpt_keys(ckpt, new_head_name='downstream_head', head_to_keep='downstream_head1', head_to_discard='downstream_head2')
            self.net.load_state_dict(ckpt["model"], strict=False)
            del ckpt
        elif isinstance(self.net, STream3R):
            log.info(f"Loading pretrained weights from {self.pretrained}")
            
            # Support both .pt and .safetensors formats
            if self.pretrained.endswith('.safetensors'):
                log.info(f"⚠️ Attempting to load from .safetensors ⚠️")
                from safetensors.torch import load_file
                checkpoint = load_file(self.pretrained)
            else:
                checkpoint = torch.load(self.pretrained)
            
            missing_keys, unexpected_keys = self.net.load_state_dict(checkpoint, strict=False)
            log.info(f"Missing keys: {missing_keys}")
            log.info(f"Unexpected keys: {unexpected_keys}")
```

However, using vggt base model (which is not a .safetensor) also comes with the issue:

```
# @package _global_

defaults:
  - override /model: stream3r

seed: 42

data:
  data_scaling: 1.0
  data_root: /workspace/data/mvs-synth/processed_GTAV_540
  num_views: 24
  allow_repeat: false
  n_corres_train: 0

  resolution:
    - [518, 392]
    - [518, 378]
    - [518, 336]
    - [518, 294]
    - [518, 252]
    - [518, 210]
    - [518, 140]
    - [378, 518]
    - [336, 518]
    - [294, 518]
    - [252, 518]
    - [224, 224]

  data_module:
    _target_: stream3r.data.multiview_dust3r_datamodule.MultiViewDUSt3RDataModule
    pin_memory: true
    num_workers: 16
    num_workers_val: 1
    batch_size_per_device: 1
    batch_size_per_device_val: 1

    train_datasets:
      - 1400 @ MVS_Synth_Multi(allow_repeat=${data.allow_repeat}, split='train', ROOT="${data.data_root}", aug_crop=16, resolution=${data.resolution}, transform=ColorJitter, num_views=${data.num_views}, n_corres=${data.n_corres_train})

model:
  pretrained: /workspace/models/vggt/model.pt
  net:
    freeze: encoder

  scheduler:
    warmup_start_lr: 1e-8
    warmup_epochs: 1

  train_criterion:
    _target_: stream3r.loss.losses.CausalLoss
    gradient_loss: grad
    is_metric: false

  validation_criterion:
    _target_: stream3r.loss.losses.CausalLoss
    gradient_loss: grad
    is_metric: false

  optimizer:
    _target_: torch.optim.AdamW
    _partial_: true
    lr: 1e-7  # Lower LR when training encoder (10× lower than paper)
    betas:
      - 0.9
      - 0.95
    weight_decay: 0.05

trainer:
  devices: auto
  max_epochs: 50
  accumulate_grad_batches: 4
  strategy:
    _target_: lightning.pytorch.strategies.DeepSpeedStrategy
    timeout:
      _target_: datetime.timedelta
      minutes: 80
  plugins: null
  precision: bf16-mixed
  log_every_n_steps: 20
  limit_val_batches: 0
  val_check_interval: 2

callbacks:
  model_checkpoint:
    every_n_train_steps: null
    every_n_epochs: 1
    save_top_k: 5
    filename: "{epoch:03d}-{step:08d}"
    save_last: false
    monitor: "train/loss"
    mode: "min"
  early_stopping:
    monitor: "train/loss"
```

By the way, I'm also using 8xA100 80GB trying to replicate your setup as much as possible (no slurm tho, single VM). I'm happy to share the dataset, or any other logs that could be helpful! Thanks again 🙌
